### PR TITLE
Reset the cache-dir when a component is added to HscEnv

### LINKED ISSUE #13: Multi component: Interface files get our of sync
If you generate interface files with one component loaded and then load another component you can get confusing error messages which don't resolve until you delete `~/.cache/ghcide`. For example 

https://github.com/haskell/haskell-language-server/issues/23#issuecomment-623444028

We need to make sure that the hash for the interface files cache takes into account all the components in the current session rather than just a hash of the options for one component.

### LINKED ISSUE #15: Multi-component: Prepend package-id to cache directory name
Or some other identifier which allows a user to see which projects have caches.

Assigning to @fendor as he is already working in this area.

### PR DESCRIPTION
**Pull request by fendor**
Also, adds a prefix for the cache-dir: InstalledUnitId.

closes #15
closes #13

### CODE REVIEW

File: exe/Main.hs [hunk 1]
```diff
@@ -287,16 +288,18 @@ loadSession dir = do
...
                 -- components in this HscEnv
-                rearrange (uid, (df, uids), ts, cfp, di) = (uid, (df, uids, ts, cfp, di))
-                do_one (uid,df, ts, cfp, di) = rearrange (uid, removeInplacePackages inplace df, ts, cfp, di)
+                rearrange (uid, (df, uids), ts, cfp, opts, di) = (uid, (df, uids, ts, cfp, opts, di))
+                do_one (uid,df, ts, cfp, opts, di) = rearrange (uid, removeInplacePackages inplace df, ts, cfp, opts, di)
                 -- All deps, but without any packages which are also loaded
                 -- into memory
-                new_deps' = map do_one new_deps
+            new_deps' <- forM new_deps $ \(uid, df_, targets, cfp, opts, di) -> do
+                df <- setCacheDir (map show inplace) opts df_
```
Comment (fendor) [inline exe/Main.hs]: `map show inplace` returns sometimes `[main, main]`, or just `[main]`, but `main` could be any executable in the package.
Investigating whether the `ComponentId` is more unique.

**Comment by mpickering** [comment]
This patch looks as I imagined it. We really need to refactor that 6-tuple into a record now!

File: exe/Main.hs [hunk 2]
```diff
@@ -430,8 +433,13 @@ loadSession dir = do
       return opts
 
 
-
-
+setCacheDir :: MonadIO m => [String] -> ComponentOptions -> DynFlags -> m DynFlags
+setCacheDir hscComponents comps dflags = do
+    let installedUnitId = show $ thisInstalledUnitId dflags
+    cacheDir <- liftIO $ getCacheDir installedUnitId (hscComponents ++ componentOptions comps)
```
Comment (mpickering) [inline exe/Main.hs]: You should sort hscComponents so the order they are opened doesn't matter.

**Comment by mpickering** [comment]
We discussed in IRC about what needs to happen for this patch to be finished.

* A comment about the invariants which need to be preserved about the cache directories and about why multiple mains are not an issue.
* Try to use just the filtered out inplace packages from the options in the salt rather than all components currently loaded.

**Comment by fendor** [comment]
Indeed and I am on it.

**Comment by mpickering** [comment]
> pretty sure it does.
> In my logs:
> 
> ```
> "Making new HscEnv[\"main\"]"
> ```
> 
> Which is from:
> 
> ```haskell
>             print ("Making new HscEnv" ++ (show inplace))
> ```

Right it does, but it might as well not as `removeInplacePackage` will never match on `main` as no components will ever depend on `main`. My idea instead was to look at which components `removeInplacePackage` actually removed and use those as the salt.

**Comment by fendor** [comment]
Ready to merge, in my opinion.
cc @mpickering

**Comment by mpickering** [comment]
So the conclusion is it fixes some issues but there are still problems?

**Comment by fendor** [comment]
for this pr, yes, but I will continue to venture into the problem on Cabal. 
For future reference this error https://gist.github.com/fendor/93d8d7685a1c09371e6e2f1ed09bd26b
can be produced by opening cabal-install, followed by Cabal and then reopening only cabal-install component.

**Comment by mpickering** [comment]
Would be good if we could reproduce it on some smaller projects other than GHC and Cabal..

**Comment by fendor** [comment]
I will provide a minimal reproducible project. I have a good idea how to reproduce it.

**Comment by fendor** [comment]
A pretty small example: https://github.com/fendor/ghcide-bad-interface-files

**Pull request merged**